components/chapters/LoginModal.tsx

```diff
@@ -0,0 +1,112 @@
+      left: '50%',
+      right: 'auto',
+      bottom: 'auto',
+      marginRight: '-50%',
+      transform: 'translate(-50%, -50%)',
+      backgroundColor: '#32455D',
+      color: '#fff',
+      width: 550,
+    },
+  }
+
+  const [loaded, setLoaded] = useState(false);
+  const [userPrivateKey, setPrivateKey] = useState('');
+
+  function loadProgress() {
+    const privateKey = Secp256k1.uint256(Buffer.from(userPrivateKey, 'hex'), 16)
+    const publicKey = Secp256k1.generatePublicKeyFromPrivateKeyData(privateKey);
+
+    let user: any = window.localStorage.getItem('user')
+    if (user) {
+      user = JSON.parse(user)
+
+      if (user.publicKey.x === publicKey.x && user.publicKey.y === publicKey.y) {
+        window.localStorage.setItem('loggedIn', "true");
+        setLoaded(true);
+      } else {
+        alert('Invalid private key')
+      }
+    }
+  }
+
+  return (
+    <Modal
+      isOpen={props.open}
+      style={customStyles}
+      contentLabel="Login Modal"
+    >
+      <div className="float-right flex justify-end">
+        <button onClick={props.onClose}>
+          <svg
+            xmlns="http://www.w3.org/2000/svg"
+            className="h-6 w-6"
+            fill="none"
+            viewBox="0 0 24 24"
+            stroke="currentColor"
+          >
+            <path
+              strokeLinecap="round"
+              strokeLinejoin="round"
+              strokeWidth={2}
+              d="M6 18L18 6M6 6l12 12"
+            />
+          </svg>
+        </button>
+      </div>
+
+      <div className="p-4">
+      {!loaded ? (
+        <>
+          <h2 className="mb-4 text-3xl font-bold">Load your progress</h2>
+          <p className="mb-5 text-xl">
+            Enter your personal code below to restore your chapter and challenge progress, as well as your avatar. <a href="#" className='text-white underline'>No code yet?</a>
+          </p>
+
+          <div className="flex flex-col justify-between">
+            <div className="flex w-full">
+              <input
+                className="bg-transparent outline-none text-white border-dotted border-white border-2 p-1 w-full text-lg"
+                type="text"
+                placeholder="Enter your code"
+                value={userPrivateKey}
+                onChange={(e) => setPrivateKey(e.target.value)}
+              />
+            </div>
+            <div className="flex w-full mt-4">
+              <button onClick={loadProgress} className="border-white border-2 w-full p-1 text-xl">Load my progress</button>
+            </div>
+          </div>
+        </>
+      ) : (
+        <>
+          <Image
+            alt="avatar"
+            src="/assets/avatars/1.png"
+            width={80}
+            height={80}
+            className="rounded-full mb-2"
+          />
+          <h2 className="mb-2 text-3xl font-bold">Progress loaded</h2>
+          <p className="mb-5 text-xl">
+            {"You're all set to continue with Chapter 3, Challenge 2."}
```

Comment:
How can we determine the users progress? Since we also need to show this on the chapters overview page (and probably elsewhere), is it best to create a utility class for this?

Reply:
I've updated it to show the user's progress.

![image](https://user-images.githubusercontent.com/22493292/208455595-e4f29d7a-5068-43f9-8b57-801e3d925e62.png)
But I don't think this is ideal, we should probably have a utility class for this in deed.